File: ruby/test/receipt_base64_test.rb
# frozen_string_literal: true

require_relative "helper"

# The spellings Receipt.decode_base64 must accept and refuse are the
# decodeBase64 groups of fixtures/cases.json, which conformance_test.rb runs
# against both the receipt-data and the x5c decoder. What stays here is what a
# shared vector cannot hold: a string that is not UTF-8 at all, which JSON text
# cannot carry, and why neither of Ruby's own decoders is the rule alone.
class ReceiptBase64Test < Minitest::Test
  Receipt = ApplePurchaseReceiptVerifier::Receipt
  VerificationError = ApplePurchaseReceiptVerifier::VerificationError
  Reason = ApplePurchaseReceiptVerifier::Reason

  def test_a_string_that_is_not_utf8_is_invalid_receipt_format
    text = "QU\xffD".b
    error = assert_raises(VerificationError, text.inspect) { Receipt.decode_base64(text) }
    assert_equal Reason::INVALID_RECEIPT_FORMAT, error.reason
    assert_nil Receipt.decode_canonical_base64(text)
  end

  # Why the shape is checked by hand: Ruby's strict decoder refuses the
  # trailing bits Apple accepts, and its lenient one skips what it does not
  # know, so neither alone is the rule.
  def test_neither_ruby_decoder_alone_is_the_rule
    assert_raises(ArgumentError) { "QR==".unpack1("m0") }
    assert_equal "ABC", "QU JD\n".unpack1("m")
  end
end

File: java/src/test/java/io/github/emindeniz99/applepurchasereceiptverifier/receipt/ReceiptBase64Test.java
package io.github.emindeniz99.applepurchasereceiptverifier.receipt;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import io.github.emindeniz99.applepurchasereceiptverifier.VerificationException;
import io.github.emindeniz99.applepurchasereceiptverifier.VerificationException.Reason;
import org.junit.jupiter.api.Test;

/**
 * The spellings {@link ReceiptBase64#decode} must accept and refuse are the
 * decodeBase64 groups of {@code fixtures/cases.json}, which
 * {@code ConformanceCasesTest} runs against it and against the x5c decoder.
 * What stays here is the one input a JSON vector cannot hold: {@code null}.
 */
class ReceiptBase64Test {

    @Test
    void nullIsInvalidReceiptFormat() {
        VerificationException thrown = assertThrows(VerificationException.class, () -> ReceiptBase64.decode(null));
        assertEquals(Reason.INVALID_RECEIPT_FORMAT, thrown.reason());
    }
}
